### PR TITLE
LSTM improvements

### LINKED ISSUE #524: Reloading reLSTM edge case
Reloading reLSTM into kernel where it's already defined results in reuse error. Either handle this explicitly, or come up with a work around.

### LINKED ISSUE #575: Add TagLSTM from SL experiments

### PR DESCRIPTION
**Pull request by henryre**
Submitted as PR due to "divine inspiration"

* Adds TagLSTM for unary candidates
* Refactors contrib and LSTM code, adding an LSTMBase class
* Bidirectionalizes LSTMs
* Adds option for attention
* Adds in-training dev set eval
* Fixes TF scoping bugs with LSTM
* Fixes TF random seeding bugs with LSTM

Fixes #575 
Fixes #524

### CODE REVIEW

**Comment by ajratner** [comment]
Will do first thing tomorrow AM!

File: snorkel/contrib/lstm/lstm_base.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        s = self.seed
+        s1, s2, s3, s4 = [None] * 4 if s is None else [s+i for i in range(4)]
+        # Embedding layer
+        emb_var   = tf.Variable(self._embedding_init(s1))
+        embedding = tf.concat([tf.zeros([1, self.dim]), emb_var], axis=0)
+        inputs    = tf.nn.embedding_lookup(embedding, self.sentences)
+        # Build RNN graph
+        batch_size = tf.shape(self.sentences)[0]
+        rand_name  = "LSTM_{0}".format(random.randint(0, 1e12)) # Obscene hack
+        init = tf.contrib.layers.xavier_initializer(seed=s2)
```
Comment (ajratner) [inline snorkel/contrib/lstm/lstm_base.py]: What is a `Xavier_initializer`?  I feel like outside of building Jarvis 5.0 in IronMan 4, this kind of name warrants a brief explanation...

File: snorkel/contrib/lstm/lstm_base.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        with tf.variable_scope(rand_name, reuse=False, initializer=init):
+            # Build LSTM cells
```
Comment (ajratner) [inline snorkel/contrib/lstm/lstm_base.py]: Should clearly delineate that this is a Bi-directional LSTM in docstring

File: snorkel/contrib/lstm/lstm_base.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+            fw_cell = rnn.BasicLSTMCell(self.dim)
+            bw_cell = rnn.BasicLSTMCell(self.dim)
```
Comment (ajratner) [inline snorkel/contrib/lstm/lstm_base.py]: Make the cell type an optional arg?

File: snorkel/contrib/lstm/lstm_base.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+            initial_state_bw = bw_cell.zero_state(batch_size, tf.float32)
+            rnn_out, _ = tf.nn.bidirectional_dynamic_rnn(
+                fw_cell, bw_cell, inputs,
+                sequence_length=self.sentence_lengths,
+                initial_state_fw=initial_state_fw,
+                initial_state_bw=initial_state_bw,
+                time_major=False               
+            )
+        # Get potentials
+        potentials = get_bi_rnn_output(rnn_out, self.dim, self.sentence_lengths)
```
Comment (ajratner) [inline snorkel/contrib/lstm/lstm_base.py]: Why is this not a standard fn?

File: snorkel/contrib/lstm/lstm_base.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        potentials_dropout = tf.nn.dropout(potentials, self.keep_prob, seed=s3)
+        W = tf.Variable(tf.random_normal((2*self.dim, 1), stddev=SD, seed=s4))
+        b = tf.Variable(0., dtype=tf.float32)
+        h_dropout = tf.squeeze(tf.matmul(potentials_dropout, W)) + b
+        # Noise-aware loss
+        self.loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(
+            labels=self.train_marginals, logits=h_dropout
+        ))
+        # Backprop trainer
+        self.train_fn = tf.train.AdamOptimizer(self.lr).minimize(self.loss)
```
Comment (ajratner) [inline snorkel/contrib/lstm/lstm_base.py]: Make this another opt arg?

File: snorkel/contrib/lstm/utils.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+def get_bi_rnn_output(output, dim, lengths):
+    c_output   = tf.concat(output, 2)
+    batch_size = tf.shape(c_output)[0]
+    max_length = tf.shape(c_output)[1]
+    index      = tf.range(0, batch_size) * max_length + (lengths - 1)
+    flat       = tf.reshape(c_output, [-1, 2 * dim])
+    return tf.gather(flat, index)
+
+def f1_score(marginals, labels, b=0.5):
```
Comment (ajratner) [inline snorkel/contrib/lstm/utils.py]: This is already in `Scorer` class?

File: snorkel/contrib/lstm/lstm_base.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+import random
+import tensorflow as tf
+import tensorflow.contrib.rnn as rnn
+
+from snorkel.learning import LabelBalancer, TFNoiseAwareModel
+from utils import f1_score, get_bi_rnn_output, SymbolTable
+from time import time
+
+
+SD = 0.1
```
Comment (henryre) [inline snorkel/contrib/lstm/lstm_base.py]: It's actually Standard Deepness units. We're not doing so well :(

File: snorkel/contrib/lstm/lstm_base.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+
+
+class LSTMBase(TFNoiseAwareModel):
+
+    representation = True
+
+    def __init__(self, save_file=None, name='LSTMBase', seed=None, n_threads=4):
+        """Base class for LSTM"""
+        # Define metadata
+        self.mx_len    = None # Max sentence length
```
Comment (henryre) [inline snorkel/contrib/lstm/lstm_base.py]: It is

File: snorkel/contrib/lstm/lstm_base.py [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+        self.dim       = None # Embedding dimension
+        self.n_v       = None # Vocabulary size
+        self.lr        = None # Learning rate
+        self.attn      = None # Attention window
```
Comment (henryre) [inline snorkel/contrib/lstm/lstm_base.py]: Ya

File: snorkel/contrib/lstm/lstm_base.py [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+        # Super constructor
+        super(LSTMBase, self).__init__(
+            n_threads=n_threads, save_file=save_file, name=name
+        )
+
+    def _preprocess_data(self, candidates, extend):
+        raise NotImplementedError()
+
+    def _make_tensor(self, x):
+        """Construct input tensor with padding"""
```
Comment (henryre) [inline snorkel/contrib/lstm/lstm_base.py]: Agreed

File: snorkel/contrib/lstm/lstm_base.py [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+        batch_size = len(x)
+        x_batch    = np.zeros((batch_size, self.mx_len), dtype=np.int32)
+        len_batch  = np.zeros(batch_size, dtype=np.int32)
+        for j, token_ids in enumerate(x):
+            t               = min(len(token_ids), self.mx_len)
```
Comment (henryre) [inline snorkel/contrib/lstm/lstm_base.py]: Yeah punted on this, but will add an error in the preprocessing steps of the subclasses

File: snorkel/contrib/lstm/lstm_base.py (hunk 1 above)
Comment (henryre) [inline snorkel/contrib/lstm/lstm_base.py]: Get deep bro. This is pretty standard for deep nets, probably ok to leave without comment.

http://jmlr.org/proceedings/papers/v9/glorot10a/glorot10a.pdf

File: snorkel/contrib/lstm/lstm_base.py (hunk 3 above)
Comment (henryre) [inline snorkel/contrib/lstm/lstm_base.py]: Yeah, was considering moving everything to `rnn` anyways

File: snorkel/contrib/lstm/lstm_base.py (hunk 4 above)
Comment (henryre) [inline snorkel/contrib/lstm/lstm_base.py]: In TF? No idea

File: snorkel/contrib/lstm/utils.py (hunk 6 above)
Comment (henryre) [inline snorkel/contrib/lstm/utils.py]: Lol it's actually not. How many F1 functions can we make?

**Comment by henryre** [comment]
@ajratner plz approve otherwise will dismiss your review (side note: need to add subgithub to urbandictionary cf. subtweet)

**Pull request merged**